browser-test/test.endtoend.ts

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright 2019 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import * as assert from 'assert';
+import {expect} from 'chai';
+import * as EchoClient from './echo_client';
+
+function sleep(timeout) {
+  return new Promise(resolve => {
+    setTimeout(resolve, timeout);
+  });
+}
+
+describe('Run tests against gRPC server', () => {
+  const authStub = {
+    getRequestHeaders() {
+      return {Authorization: 'Bearer SOME_TOKEN'};
+    },
+  };
+
+  const opts = {
+    auth: authStub,
+  };
+
+  const client = new EchoClient(opts);
+
+  before(async function() {
+    this.timeout(40000);
+
+    const request = {
+      content: 'test',
+    };
+    const MAX_RETRIES = 20;
+    const TIMEOUT = 1000;
+    let retryCount = 0;
+    // Making sure server is up before starting tests.
+    for (; retryCount < MAX_RETRIES; ++retryCount) {
+      try {
+        await client.echo(request);
+      } catch (err) {
+        console.log('Still waiting for server...');
```

Comment:
I usually try to avoid having console messages in tests, as they can interfere with the test runner output; I've been known to use the module [debug](https://www.npmjs.com/package/debug) for this sort of thing.

Reply:
I removed `console.log` for now. Seems like we can live without it.